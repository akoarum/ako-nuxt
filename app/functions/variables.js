export const VALIDATES = {
  EMPTY: {
    message: '入力してください。'
  },
  NUMBER: {
    regex: /^[0-9]+?$/,
    message: '数字で入力してください。'
  },
  EMAIL: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスの形式になっていません。'
  },
  TEL: {
    regex: /^[0-9]{2,4}-?[0-9]{2,4}-?[0-9]{4}$/,
    message: '電話番号の形式になっていません。'
  },
  URL: {
    regex: /^http(s):\/\/.+?\.+?$/,
    message: 'URLの形式になっていません。'
  }
}
