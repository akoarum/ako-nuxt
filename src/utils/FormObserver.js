export default class FormValidator {
  constructor(names = []) {
    this._items = this._createItems(names)
  }

  /**
   * @return {Boolean}
   */
  get btnDisabled() {
    for (const name in this._items) {
      if (this._items[name].error || !this._items[name].dirty) return true
    }
    return false
  }

  /**
   * itemオブジェクトの生成
   * @param {Array} names name配列
   */
  _createItems(names) {
    const items = {}

    names.forEach(item => {
      items[item] = {
        error: false,
        dirty: false
      }
    })

    return items
  }

  /**
   * itemの更新
   * @param {String}  name  name値
   * @param {String}  prop  更新するプロパティ名
   * @param {Boolean} value 値
   */
  updateItem(name, prop, value) {
    this._items[name][prop] = value
  }
}
