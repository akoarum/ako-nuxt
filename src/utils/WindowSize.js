export default class WindowSize {
  constructor(delay = 80) {
    this._width = 0
    this._height = 0
    this._delay = delay
    this._cullTime = null
    this._initialized = false
    this._handler = this._handleResize.bind(this)
  }

  get width() {
    return this._width
  }

  get height() {
    return this._height
  }

  get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  initialize() {
    if (this._initialized) return this
    this._updateState()
    window.addEventListener('resize', this._handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  _updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  _handleResize() {
    clearTimeout(this._cullTime)
    this._cullTime = setTimeout(() => {
      this._updateState()
    }, this._delay)
  }

  /**
   * 解放する
   */
  destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this._handler)
    this._initialized = false
    return this
  }
}
