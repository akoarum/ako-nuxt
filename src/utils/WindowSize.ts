export default class WindowSize {
  private _width: number = 0
  private _height: number = 0
  private delay: number = 80
  private cullTime = null
  private _initialized: boolean = false
  private handler = this.handleResize.bind(this)

  constructor(delay: number = 80) {
    this.delay = delay
  }

  public get width() {
    return this._width
  }

  public get height() {
    return this._height
  }

  public get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  public initialize() {
    if (this._initialized) return this
    this.updateState()
    window.addEventListener('resize', this.handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  public updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  private handleResize() {
    clearTimeout(this.cullTime)
    this.cullTime = setTimeout(() => {
      this.updateState()
    }, this.delay)
  }

  /**
   * 解放する
   */
  public destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this.handler)
    this._initialized = false
    return this
  }
}
