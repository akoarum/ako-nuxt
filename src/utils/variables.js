export const VALIDATES = {
  required: '入力してください。',
  selected: '選択してください。',
  file: 'ファイルをアップロードしてください。',
  tel: {
    regex: /^\d{2,4}-?\d{4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
