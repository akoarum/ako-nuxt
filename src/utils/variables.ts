import { ValidatePattern } from '~/types/models'

export const VALIDATE_PATTERNS: { [key: string]: ValidatePattern } = {
  required: {
    message: '入力してください。'
  },
  selected: {
    message: '選択してください。'
  },
  tel: {
    regex: /^\d{2,5}-?\d{1,4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
