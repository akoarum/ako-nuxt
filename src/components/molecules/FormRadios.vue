<template>
  <div class="formRadios">
    <ul class="formRadios__list">
      <li v-for="select in selects" :key="select.id" class="formRadios__item">
        <VRadio
          v-model="model"
          v-bind="$attrs"
          :name="name"
          :value="select.value"
          :aria-invalid="String(!!error)"
          :aria-describedby="error ? `error-${name}` : false"
          @change.once="updateDirty"
        >
          {{ select.label }}
        </VRadio>
      </li>
    </ul>
    <VTexts
      v-if="error"
      :id="`error-${name}`"
      role="alert"
      aria-live="assertive"
      caution
      small
      class="formRadios__error"
    >
      <span>{{ error }}</span>
    </VTexts>
  </div>
</template>

<script>
import { VALIDATES } from '~/utils/variables'
import { VRadio, VTexts } from '~/components/atoms'

export default {
  components: { VRadio, VTexts },
  inheritAttrs: false,
  model: {
    event: 'change'
  },
  props: {
    name: { type: String, required: true },
    selects: { type: Array, required: true },
    value: { type: String },
    required: { type: Boolean }
  },
  data() {
    return {
      error: '',
      isDirty: !!this.value || !this.required
    }
  },
  computed: {
    model: {
      set(value) {
        this.$emit('change', value)
        this.error = this.validate(value)
      },
      get() {
        return this.value
      }
    }
  },
  watch: {
    error(value) {
      this.$emit('error', { name: this.name, error: !!value })
    },
    isDirty() {
      this.$emit('dirty', this.name)
    }
  },
  methods: {
    validate(value) {
      if (!this.required) return ''
      if (!this.isDirty) return ''
      if (!value) return VALIDATES.selected
      return ''
    },
    updateDirty() {
      this.isDirty = true
    }
  }
}
</script>

<style lang="scss" scoped>
.formRadios__item + .formRadios__item {
  margin-top: 12px;
}
.formRadios__error {
  @include media-min {
    margin-top: 6px;
  }
  @include media-max {
    margin-top: 8px;
  }
}
</style>
