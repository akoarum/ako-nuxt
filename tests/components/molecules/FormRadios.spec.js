import { shallowMount } from '@vue/test-utils'
import FormRadios from '~/components/molecules/FormRadios'
import { VALIDATES } from '~/utils/variables'

describe('FormRadios', () => {
  let wrapper
  const selects = [
    { id: 1, label: 'ラベル1', value: '1' },
    { id: 2, label: 'ラベル2', value: '2' },
    { id: 3, label: 'ラベル3', value: '3' },
    { id: 4, label: 'ラベル4', value: '4' }
  ]

  beforeEach(() => {
    wrapper = shallowMount(FormRadios, {
      propsData: {
        name: 'name',
        value: '',
        selects
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('validate', () => {
      it('requiredでisDirtyでない場合は空文字列を返す', () => {
        expect(wrapper.vm.validate('1')).toBe('')
      })

      it('requiredかつisDirtyでvalueが空ならselectedエラーを返す', () => {
        wrapper.setProps({ required: true })
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('')).toBe(VALIDATES.selected)
      })

      it('valueが空でない場合は空文字列を返す', () => {
        wrapper.setProps({ isDirty: true })
        expect(wrapper.vm.validate('1')).toBe('')
      })
    })

    describe('updateDirty', () => {
      it('isDirtyをtrueにする', () => {
        wrapper.vm.updateDirty()
        expect(wrapper.vm.isDirty).toBe(true)
      })
    })
  })
})
