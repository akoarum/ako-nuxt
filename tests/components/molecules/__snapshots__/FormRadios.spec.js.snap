// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRadios initialized snapshot 1`] = `
<div
  class="formRadios"
>
  <ul
    class="formRadios__list"
  >
    <li
      class="formRadios__item"
    >
      <vradio-stub
        aria-invalid="false"
        name="name"
        value=""
      >
        
        ラベル1
      
      </vradio-stub>
    </li>
    <li
      class="formRadios__item"
    >
      <vradio-stub
        aria-invalid="false"
        name="name"
        value=""
      >
        
        ラベル2
      
      </vradio-stub>
    </li>
    <li
      class="formRadios__item"
    >
      <vradio-stub
        aria-invalid="false"
        name="name"
        value=""
      >
        
        ラベル3
      
      </vradio-stub>
    </li>
    <li
      class="formRadios__item"
    >
      <vradio-stub
        aria-invalid="false"
        name="name"
        value=""
      >
        
        ラベル4
      
      </vradio-stub>
    </li>
  </ul>
   
  <!---->
</div>
`;
